Add Concourse CI chart

frodenas opened the pull request:
This PR add a Helm chart for [Concourse](https://concourse.ci/).

It requires Kubernetes >= 1.5, as it uses `StatefulSets`. Using persistent storage is optional (enabled by default, but configurable by properties).

frodenas committed:
Add Concourse CI chart

enisoc commented on incubator/concourse/Chart.yaml:
```diff
@@ -0,0 +1,15 @@
+name: concourse
+version: 0.0.1
```
Usually the first version is 0.1.0.

lachie83 replied:
Agreed. Please bump

enisoc commented on incubator/concourse/README.md:
````diff
@@ -0,0 +1,243 @@
+$ ssh-keygen -t rsa -f concourse-keys/worker_key -N '' -C concourse
+```
+
+And update the `values.yaml` file with the generated keys:
````
You can keep the keys in separate files in the chart dir, and then [include them from your template files](https://github.com/kubernetes/helm/blob/master/docs/charts.md#predefined-values) with something like this:

```yaml
data:
  hostKey: |-
{{ .Files.Get "host_key" | indent 4}}
```

That way users don't have to paste the file contents directly into `values.yaml`.

frodenas replied:
@enisoc Users deploying this chart in production are encouraged to provide their own keys (as outlined at the `README`). Having default keys is really convenient, but not secure. I don't like the actual behavior, agreed, but I didn't find any elegant solution.

If I use the `.Files.Get` approach at the `configMap`, that means that if the user wants to provide keys, it needs to clone the repo and add those keys into the cloned chart directory. `.Files.Get` does NOT work inside the `values.yml`, so nothing I can do here. 

A proper solution might be what is outlined at https://github.com/kubernetes/helm/issues/1754. Having the ability to provide files using the command line with help here a lot.

I'm a newbie, so correct me if I'm wrong, or if there's any other solution to this.

viglesiasce replied:
You could do something like the ghost and gitlab-ce charts where they dont actually deploy pods until a value is set. In the NOTES you could then print out that they need to set values as follows:

$ ssh-keygen ....
$ helm upgrade -n my-release --set privateKey=`cat id_rsa` --set publicKey=`cat id_rsa.pub`

frodenas replied:
@viglesiasce Wouldn't it be easier if I just update the README? So instead of `And update the values.yaml file with the generated keys` will be something like `And provide the generated keys by using helm install --set privateKey= cat id_rsa  --set publicKey= cat id_rsa.pub`?

frodenas replied:
I just updated the README to show how users can set the keys using the `--set` command line flag.

enisoc replied:
@frodenas I see what you mean about `.Files.Get` not working here. Your two alternatives look good to me then. A third option might be to help them generate a secondary values file with a command they can copy-paste:

```console
$ cat >keys.yaml <<EOF
concourse:
  hostKey: |-
$(sed 's/^/    /' concourse-keys/host_key)
  hostKeyPub: |-
$(sed 's/^/    /' concourse-keys/host_key.pub)
  sessionSigningKey: |-
$(sed 's/^/    /' concourse-keys/session_signing_key)
  sessionSigningKeyPub: |-
$(sed 's/^/    /' concourse-keys/session_signing_key.pub)
  workerKey: |-
$(sed 's/^/    /' concourse-keys/worker_key)
  workerKeyPub: |-
$(sed 's/^/    /' concourse-keys/worker_key.pub)
EOF
$ helm install ... -f values.yaml -f keys.yaml
```

enisoc commented on incubator/concourse/templates/worker-statefulset.yaml:
```diff
@@ -0,0 +1,116 @@
+      labels:
+        app: {{ template "worker.fullname" . }}
+      annotations:
+        pod.alpha.kubernetes.io/initialized: "true"
```
This annotation is no longer necessary for StatefulSet. It was only needed for the old incarnation called PetSet.

enisoc commented on incubator/concourse/templates/worker-statefulset.yaml:
```diff
@@ -0,0 +1,116 @@
+    - metadata:
+        name: concourse-work-dir
+        annotations:
+          volume.alpha.kubernetes.io/storage-class: {{ .Values.persistence.worker.class | quote }}
```
The `volume.alpha` version of the `storage-class` annotation doesn't actually use its value. To support both the new behavior (which requires a value) and the old behavior, we recommend a pattern like this:

https://github.com/kubernetes/charts/pull/361/commits/95626d44db84935804aa029a573c8c75be2090f3

lachie83 commented:
@frodenas Thanks for the contrib!

lachie83 commented:
@enisoc fantastic! Thank you very much for jumping in with the review. Very much appreciated

lachie83 commented on incubator/concourse/templates/_helpers.tpl:
```diff
@@ -0,0 +1,35 @@
+*/}}
+{{- define "postgresql.fullname" -}}
+{{- $name := default "postgresql" .Values.postgresql.nameOverride -}}
+{{- printf "%s-%s" .Release.Name $name | trunc 24 | trimSuffix "-" -}}
```
Any reason this isn't 63 as well?

frodenas replied:
@lachie83 `postgresql` is a dependency chart (and is using 24). If I use 63 on my `_helpers.tpl`, will it be propagated to the postgresql deployment? I will do a quick test to check the behavior.

frodenas replied:
@lachie83 I did a quick test, and doesn't work if I change this to 63. The reason is that the `postgresql` chart is using a [not scoped fullname](https://github.com/kubernetes/charts/blob/master/stable/postgresql/templates/_helpers.tpl). 

So or I change my `_helpers.tpl` to use `fullname` instead of `postgresql.fullname` (and this will be very confusing), or I update the `postgresql` chart to use a scoped fullname.

lachie83 replied:
Appreciate you testing. Let's leave as-is

lachie83 commented:
@k8s-bot ok to test

viglesiasce commented on incubator/concourse/values.yaml:
```diff
@@ -0,0 +1,375 @@
+    ## For minikube, set this to NodePort, elsewhere use LoadBalancer.
+    ## ref: https://kubernetes.io/docs/user-guide/services/#publishing-services---service-types
+    ##
+    type: NodePort
```
Make this a ClusterIP service and use port forwarding by default since the out of the box config uses static credentials.

viglesiasce reviewed (changes requested):
This worked really well for me. Thanks @frodenas!!

Only one quibble on the default service type for the web.

viglesiasce reviewed (commented):
I would also move this to the stable directory since StatefulSet is now beta

frodenas commented:
@viglesiasce I changed the default service type to `ClusterIP` and moved the chart from `incubator` to `stable`. Tests failed due to something unrelated: `/src/test/changed.sh: line 43: GOOGLE_APPLICATION_CREDENTIALS: unbound variable`. Can you please check the e2e tests logs? Seems that variable is not set.

frodenas commented:
@viglesiasce Pushed it again and now the e2e tests succeed.

Commit:
concourse: Add namespace to port-forward
